sematic/ui/src/components/RunPanel.tsx

```diff
@@ -152,3 +126,57 @@ export default function RunPanel(props: {
     </Box>
   );
 }
+
+function RunActionMenu(props: {
+  run: Run;
+  edges: Edge[];
+  resolution: Resolution;
+}) {
+  const { run, edges, resolution } = props;
+
+  const { user } = useContext(UserContext);
+
+  const { setSnackMessage } = useContext(SnackBarContext);
+
+  const onRerunClick = useCallback(() => {
+    fetchJSON({
+      url: "/api/v1/resolutions/" + run.root_id + "/rerun",
+      method: "POST",
+      body: { rerun_from: run.id },
+      apiKey: user?.api_key,
+      callback: (payload) => {},
+      setError: (error) => {
+        if (error) setSnackMessage({ message: "Failed to trigger a rerun" });
+      },
+    });
+  }, []);
+
+  const rerunEnabled = useMemo(
+    () =>
+      edges.every((edge) => !!edge.artifact_id) &&
```

Comment:
Can't find what "!!" means on the Internet. Does this mean is defined and non-null?

Reply:
It's not something special, it just means double `!`. It's a way to convert anything into a bool.
`artifact_id` is either a string or `null`. `!!null` will be `false` and `!!some_string` will be `true`.